### PR TITLE
Fixed #26517 -- Select 0 for impossible predicates

### PR DESCRIPTION
**Pull request by emulbreh**
* https://code.djangoproject.com/ticket/26517
* https://code.djangoproject.com/ticket/26433

### CODE REVIEW

File: tests/annotations/tests.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from django.core.exceptions import FieldDoesNotExist, FieldError
 from django.db.models import (
     BooleanField, CharField, Count, DateTimeField, ExpressionWrapper, F, Func,
-    IntegerField, Sum, Value,
+    IntegerField, Sum, Value, Q
```
Comment (charettes) [inline tests/annotations/tests.py]: Put `Q` before `Sum` to appease isort.

**Comment by charettes** [comment]
Buildbot, add to whitelist.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -223,7 +223,12 @@ def get_select_from_parent(klass_info):
 
         ret = []
         for col, alias in select:
-            ret.append((col, self.compile(col, select_format=True), alias))
+            try:
+                sql, params = self.compile(col, select_format=True)
+            except EmptyResultSet:
+                # Selecting a predicate that is always false:
+                sql, params = '0', ()
```
Comment (jarshwah) [inline django/db/models/sql/compiler.py]: This isn't safe unfortunately. Consider the following test results (which fail):

```
    def test_empty_expression_char(self):
        books = Book.objects.annotate(
            selected=Case(
                When(pk__in=Book.objects.none(), then=Value('Empty')),
                default=Value('Not Empty'), output_field=CharField()
            )
        )
        self.assertGreater(len(books), 0)
        self.assertEqual(books[0].selected, 'Not Empty')

    def test_empty_expression_datetime(self):
        from django.utils import timezone
        from datetime import timedelta
        now = timezone.now()
        then = now + timedelta(days=1)

        books = Book.objects.annotate(
            selected=Case(
                When(pk__in=Book.objects.none(), then=Value(now)),
                default=Value(then), output_field=DateTimeField()
            )
        )
        self.assertGreater(len(books), 0)
        self.assertEqual(books[0].selected, then)
```

Depending on whether or not there are dbconverters run, these will fail with one of the following kinds of errors:

1. `TypeError: expected string or buffer`
2. `AssertionError: 0 != datetime.datetime(2016, 5, 23, 18, 59, 26, 409777)`

For what it's worth, both of these tests fail without your patch too, except they fail during the count just like the report on the ticket says. This certainly needs fixing in some way, but we can't just use 0 as a value all of the time.

Comment (emulbreh) [inline django/db/models/sql/compiler.py]: I'm not familiar with orm internals anymore, is this problem specific to `Case`?  If so, couldn't we catch EmptyResultSet in `When` as well?

Comment (jarshwah) [inline django/db/models/sql/compiler.py]: Case expressions use Q internally which led me to writing the above tests. I think you're right that EmptyResultSet might be able to be caught directly, and then it could use the `default` argument in place of the static `'0'` you have above. The changes need to be made in tandem though.

Comment (emulbreh) [inline django/db/models/sql/compiler.py]: The same try/except fixes `When` (see last commit). I'll try to push a cleaned up commit tonight (UTC+1).

File: django/db/models/expressions.py [hunk 3]
```diff
@@ -843,13 +843,19 @@ def as_sql(self, compiler, connection, template=None, case_joiner=None, **extra_
         template_params.update(extra_context)
         case_parts = []
         sql_params = []
+        from django.db.models.sql.datastructures import EmptyResultSet
```
Comment (emulbreh) [inline django/db/models/expressions.py]: I don't see a good way to solve this import cycle. Should I move EmptyResultSet out of models.sql?

Comment (jarshwah) [inline django/db/models/expressions.py]: Since EmptyResultSet isn't necessarily related to SQL I think moving it makes sense. `django.core.exceptions` is probably a good place to move it to. The only question is whether we need to deprecate the old location, or at least provide an import in the old location.

@akaariai @loic @manfre @adamchainz do any of you have an opinion here? I imagine non-core backends are likely to import this type but I haven't checked specifically.

In any case, if the change is made, it should be in a separate commit to catch regressions in the future without impacting 26517 if it needs to be reverted.

Comment (adamchainz) [inline django/db/models/expressions.py]: I have no opinion, Django-MySQL doesn't touch `EmptyResultSet`

File: django/db/models/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+class EmptyResultSet(Exception):
+    pass
```
Comment (charettes) [inline django/db/models/exceptions.py]: This module seems unused.

**Comment by charettes** [comment]
Do you think you could add an extra `Coalesce` test as this should also solve [#26430](https://code.djangoproject.com/ticket/26430)

File: tests/annotations/tests.py [hunk 5]
```diff
@@ -148,6 +148,16 @@ def test_mixed_type_annotation_numbers(self):
         combined = int(test.pages + test.rating)
         self.assertEqual(b.combined, combined)
 
+    def test_empty_expression_annotation(self):
+        books = Book.objects.annotate(
+            selected=ExpressionWrapper(Q(pk__in=[]), output_field=BooleanField())
+        )
+        self.assertGreater(len(books), 0)
```
Comment (timgraham) [inline tests/annotations/tests.py]: Why such a vague assertion? At least you could run `Book.objects.count()` at the beginning of the test and check the count if that's the expected result. What about also verifying the purpose of the annotation -- it's not obvious to me what that purpose is.

File: django/db/models/expressions.py (hunk 3 above)
Comment (timgraham) [inline django/db/models/expressions.py]: The import usually goes at the top of the method.

**Comment by timgraham** [comment]
Should the second commit message reference ticket #26433?

File: tests/annotations/tests.py [hunk 6]
```diff
@@ -158,6 +158,16 @@ def test_empty_expression_annotation(self):
         )
         self.assertGreater(len(books), 0)
 
+    def test_empty_when_expression(self):
```
Comment (timgraham) [inline tests/annotations/tests.py]: This test probably belongs in `tests/expressions_case`.

File: django/db/models/expressions.py (hunk 3 above)
Comment (timgraham) [inline django/db/models/expressions.py]: I think exceptions in `django.core.exceptions` are mostly things that a user might need to import. I don't think `EmptyResultSet` meets that criteria, at least in my experience I never needed it in a project.

Comment (jarshwah) [inline django/db/models/expressions.py]: Good point, I didn't think of that module as being user facing. Perhaps django.db.utils would be a better place? It already contains various DatabaseError exceptions, so it wouldn't be totally out of place.

Comment (timgraham) [inline django/db/models/expressions.py]: Maybe, I don't feel strongly about the need to move it just to avoid some circular imports. On the other hand, now the `Case` needs it, is there a chance that other expressions might need it too?

Comment (jarshwah) [inline django/db/models/expressions.py]: Yes there is a chance it may be needed by future expressions. The subquery expression that @schinckel is putting together may end up needing it too. Purity is also a reason though. The error isn't really sql specific so it shouldn't be pushed down into the sql implementation - much nicer to have it somewhere in the frontend. History tells me it used to live in `django.db.models.query` until it was moved into `django.db.models.sql.query` during the queryset refactor - but I think that was just because it was used only within the sql module at that point.

Comment (timgraham) [inline django/db/models/expressions.py]: I'm fine with core.exceptions then. Docs go in docs/ref/exceptions.txt.

**Comment by emulbreh** [comment]
@charettes #26430 has similar symptoms but isn't fixed by this PR. Since the EmptyResultSet in #26430 is caught by `execute_sql()` we can't easily handle it in the aggregate functions.

**Comment by timgraham** [comment]
Could you please squash the follow up commits into the appropriate commit? Also docs for `EmpytResultSet` go in docs/ref/exceptions.txt. When finished, uncheck "patch needs improvement" on the ticket so it appears in the review queue.

**Comment by emulbreh** [comment]
@timgraham I don't think EmptyResultSet should be public api. It's not supposed to be raised or caught outside of `django.db.models`. If necessary I can add a paragraph in the docs anyway, but I'm not sure what to write.

**Comment by timgraham** [comment]
See this discussion about EmptyResultSet: https://github.com/django/django/pull/6638#discussion_r68674228

File: tests/annotations/tests.py [hunk 7]
```diff
@@ -148,6 +148,19 @@ def test_mixed_type_annotation_numbers(self):
         combined = int(test.pages + test.rating)
         self.assertEqual(b.combined, combined)
 
+    def test_empty_expression_annotation(self):
+        books = Book.objects.annotate(
+            selected=ExpressionWrapper(Q(pk__in=[]), output_field=BooleanField())
```
Comment (timgraham) [inline tests/annotations/tests.py]: @jarshwah -- is this test okay to commit despite your comment on the ticket, "ExpressionWrapper was never meant to be used to wrap up Q() objects. Q() are not really expressions themselves, they just implement some of the same API internally."?

Comment (jarshwah) [inline tests/annotations/tests.py]: I would prefer it to be removed since it's not something we've committed to supporting. The problem is that there are many people already using Q in this way rather than the longer but more correct Case/When. I'm fairly sure wrapped Qs will break in subtle ways but I haven't dug into it enough to come up with a breaking scenario.

Comment (timgraham) [inline tests/annotations/tests.py]: We'll revert the changes in `django/db/models/sql/compiler.py` as well then? This is acting as a regression test for those changes.

Comment (jarshwah) [inline tests/annotations/tests.py]: Ok, let's leave the compiler changes and the test. There are enough people using this structure for us to either break it outright or support it. We can always fix any bugs that appear with `ExpressionWrapper(Q)`.

FWIW the problems I considered were to do with relabelling aliases or resolve_expression running multiple times. Now Q is just a resolvable, not really an expression. Whatever it returns must be a genuine expression. It returns `WhereNode` which has everything that an expression should have *except* `resolve_expression` and `output_field`.

If there's a way to trigger resolve_expression to be run multiple times, perhaps by using a queryset as a subquery, then it might be possible to trigger an error. I'm just brain dumping right now, I don't expect this to be tested for this patch.

Anyway, if we're going to all this trouble to support Q() objects as fields, then why not just add an `output_field = BooleanField()` to the WhereNode and bypass ExpressionWrapper altogether?

Comment (timgraham) [inline tests/annotations/tests.py]: > "why not just add an output_field = BooleanField() to the WhereNode and bypass ExpressionWrapper altogether"`

This belongs in a new ticket?

Comment (jarshwah) [inline tests/annotations/tests.py]: Yes, I agree. Leave the test and the compiler changes in. I'll create a new ticket for full Q expression support.

**Comment by timgraham** [comment]
I made a few cosmetic updates, split the first commit into two commits, and added documentation for `EmptyResultResult` in https://github.com/django/django/pull/7028. Let me know if that looks look good, otherwise I think this is ready for commit.

Thanks Johannes!
